server/src/Core.hs

```diff
@@ -74,70 +90,116 @@ type ServerState = SubscriptionTree UUID WS.Connection
 newServerState :: ServerState
 newServerState = empty
 
-newCore :: Value -> Config -> Maybe Metrics.IcepeakMetrics -> IO Core
-newCore initialValue config metrics = do
+-- | Try to initialize the core. This loads the database and sets up the internal data structures.
+newCore :: Config -> Maybe Metrics.IcepeakMetrics -> IO (Either String Core)
+newCore config metrics = do
   let queueCapacity = fromIntegral . configQueueCapacity $ config
-  tvalue <- newTVarIO initialValue
-  tqueue <- newTBQueueIO queueCapacity
-  tupdates <- newTBQueueIO queueCapacity
-  tclients <- newMVar newServerState
-  tlogrecords <- newTBQueueIO queueCapacity
-  pure (Core tvalue tqueue tupdates tclients tlogrecords config metrics)
+  -- load the persistent data from disk
+  let filePath = configDataFile config
+  eitherValue <- readData filePath metrics
+  for eitherValue $ \initialValue -> do
+    -- create synchronization channels
+    tvalue <- newTVarIO initialValue
+    tdirty <- newTVarIO False
+    tqueue <- newTBQueueIO queueCapacity
+    tupdates <- newTBQueueIO queueCapacity
+    tclients <- newMVar newServerState
+    tlogrecords <- newTBQueueIO queueCapacity
+    pure (Core tvalue tdirty tqueue tupdates tclients tlogrecords config metrics)
 
 -- Tell the put handler loop, the update handler and the logger loop to quit.
 postQuit :: Core -> IO ()
 postQuit core = do
   atomically $ do
-    writeTBQueue (coreQueue core) Nothing
+    writeTBQueue (coreQueue core) Stop
     writeTBQueue (coreUpdates core) Nothing
     writeTBQueue (coreLogRecords core) Nothing
 
-enqueueOp :: Op -> Core -> IO EnqueueResult
-enqueueOp op core = atomically $ do
+-- | Try to enqueue a command. It succeeds if the queue is not full, otherwise,
+-- nothing is changed.
+tryEnqueueCommand :: Command -> Core -> IO EnqueueResult
+tryEnqueueCommand op core = atomically $ do
   isFull <- isFullTBQueue (coreQueue core)
-  unless isFull $ writeTBQueue (coreQueue core) (Just op)
+  unless isFull $ writeTBQueue (coreQueue core) op
   pure $ if isFull then Dropped else Enqueued
 
+-- | Enqueue a command. Blocks if the queue is full.
+enqueueCommand :: Command -> Core -> IO ()
+enqueueCommand cmd core = atomically $ writeTBQueue (coreQueue core) cmd
```

Comment:
When/why should this be used instead of `tryEnqueueCommand`?

Reply:
Added explanation